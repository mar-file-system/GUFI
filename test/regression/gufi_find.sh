#!/usr/bin/env bash

# This file is part of GUFI, which is part of MarFS, which is released
# under the BSD license.
#
#
# Copyright (c) 2017, Los Alamos National Security (LANS), LLC
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without modification,
# are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice, this
# list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation and/or
# other materials provided with the distribution.
#
# 3. Neither the name of the copyright holder nor the names of its contributors
# may be used to endorse or promote products derived from this software without
# specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
# OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
#
# From Los Alamos National Security, LLC:
# LA-CC-15-039
#
# Copyright (c) 2017, Los Alamos National Security, LLC All rights reserved.
# Copyright 2017. Los Alamos National Security, LLC. This software was produced
# under U.S. Government contract DE-AC52-06NA25396 for Los Alamos National
# Laboratory (LANL), which is operated by Los Alamos National Security, LLC for
# the U.S. Department of Energy. The U.S. Government has rights to use,
# reproduce, and distribute this software.  NEITHER THE GOVERNMENT NOR LOS
# ALAMOS NATIONAL SECURITY, LLC MAKES ANY WARRANTY, EXPRESS OR IMPLIED, OR
# ASSUMES ANY LIABILITY FOR THE USE OF THIS SOFTWARE.  If software is
# modified to produce derivative works, such modified software should be
# clearly marked, so as not to confuse it with the version available from
# LANL.
#
# THIS SOFTWARE IS PROVIDED BY LOS ALAMOS NATIONAL SECURITY, LLC AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
# THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL LOS ALAMOS NATIONAL SECURITY, LLC OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
# OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
# IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
# OF SUCH DAMAGE.



set -e

ROOT="$(realpath ${BASH_SOURCE[0]})"
ROOT="$(dirname ${ROOT})"
ROOT="$(dirname ${ROOT})"
ROOT="$(dirname ${ROOT})"

# gufi_find wrapper
GUFI_FIND="${ROOT}/test/regression/gufi_find.py"

# output directories
SRCDIR="prefix"
INDEXROOT="${SRCDIR}.gufi"

source ${ROOT}/test/regression/setup.sh "${ROOT}" "${SRCDIR}" "${INDEXROOT}"

OUTPUT="gufi_find.out"

function replace() {
    echo "$@" | sed "s/${GUFI_FIND//\//\\/}/gufi_find/g; s/${INDEXROOT//\//\\/}\\///g; s/${INDEXROOT//\//\\/}/./g; s/[[:space:]]*$//g"
}

function run_unsorted() {
    replace "$ $@"
    replace "$($@)"
    echo
}

function run_sorted() {
    replace "$ $@"
    replace "$($@)" | sort
    echo
}

(
run_sorted   "${GUFI_FIND}"
run_sorted   "${GUFI_FIND} -empty"
run_sorted   "${GUFI_FIND} -type d"
run_sorted   "${GUFI_FIND} -type f"
run_sorted   "${GUFI_FIND} -type l"
run_sorted   "${GUFI_FIND} -readable"
run_sorted   "${GUFI_FIND} -writable"
run_sorted   "${GUFI_FIND} -executable"
run_sorted   "${GUFI_FIND} -mindepth 2"
run_sorted   "${GUFI_FIND} -maxdepth 0"
# don't include directories or links because their
# sizes are not consistent across filesystems
run_sorted   "${GUFI_FIND} -type f -size 1c"
run_sorted   "${GUFI_FIND} -type f -size=-1c"
run_sorted   "${GUFI_FIND} -type f -size +1c"
run_sorted   "${GUFI_FIND} -type f -size +1024c"
run_sorted   "${GUFI_FIND} -type f -size +1 -size=-3" # 512 < size < 1536
run_sorted   "${GUFI_FIND} -type f -size 2048"        # 512 * 2048 = 1MB
run_unsorted "${GUFI_FIND} -type f --smallest"
run_unsorted "${GUFI_FIND} -type f --largest"
) | tee "${OUTPUT}"

diff ${ROOT}/test/regression/gufi_find.expected "${OUTPUT}"
rm "${OUTPUT}"
