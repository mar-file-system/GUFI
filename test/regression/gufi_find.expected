$ gufi_find
.
.hidden
1KB
1MB
directory
directory/executable
directory/readonly
directory/subdirectory
directory/subdirectory/directory_symlink
directory/subdirectory/repeat_name
directory/writable
empty_file
file_symlink
leaf_directory
leaf_directory/leaf_file1
leaf_directory/leaf_file2
old_file
repeat_name
unusual, name?#

$ gufi_find -empty
empty_file

$ gufi_find -type d
.
directory
directory/subdirectory
leaf_directory

$ gufi_find -type f
.hidden
1KB
1MB
directory/executable
directory/readonly
directory/subdirectory/repeat_name
directory/writable
empty_file
leaf_directory/leaf_file1
leaf_directory/leaf_file2
old_file
repeat_name
unusual, name?#

$ gufi_find -type l
directory/subdirectory/directory_symlink
file_symlink

$ gufi_find -readable
.
.hidden
1KB
1MB
directory
directory/executable
directory/readonly
directory/subdirectory
directory/subdirectory/directory_symlink
directory/subdirectory/repeat_name
directory/writable
empty_file
file_symlink
leaf_directory
leaf_directory/leaf_file1
leaf_directory/leaf_file2
old_file
repeat_name
unusual, name?#

$ gufi_find -writable
.
.hidden
1KB
1MB
directory
directory/executable
directory/subdirectory
directory/subdirectory/directory_symlink
directory/subdirectory/repeat_name
directory/writable
empty_file
file_symlink
leaf_directory
leaf_directory/leaf_file1
leaf_directory/leaf_file2
old_file
repeat_name
unusual, name?#

$ gufi_find -executable
.
directory
directory/executable
directory/subdirectory
directory/subdirectory/directory_symlink
file_symlink
leaf_directory

$ gufi_find -mindepth 2
directory/subdirectory
directory/subdirectory/directory_symlink
directory/subdirectory/repeat_name

$ gufi_find -maxdepth 0
.
.hidden
1KB
1MB
empty_file
file_symlink
old_file
repeat_name
unusual, name?#

$ gufi_find -type f -size 1c
.hidden
directory/executable
directory/readonly
directory/subdirectory/repeat_name
directory/writable
leaf_directory/leaf_file1
leaf_directory/leaf_file2
old_file
repeat_name
unusual, name?#

$ gufi_find -type f -size=-1c
empty_file

$ gufi_find -type f -size +1c
1KB
1MB

$ gufi_find -type f -size +1024c
1MB

$ gufi_find -type f -size +1 -size=-3
1KB

$ gufi_find -type f -size 2048
1MB

$ gufi_find -type f --smallest
empty_file
.hidden
directory/executable
directory/readonly
directory/subdirectory/repeat_name
directory/writable
leaf_directory/leaf_file1
leaf_directory/leaf_file2
old_file
repeat_name
unusual, name?#
1KB
1MB

$ gufi_find -type f --largest
1MB
1KB
.hidden
directory/executable
directory/readonly
directory/subdirectory/repeat_name
directory/writable
leaf_directory/leaf_file1
leaf_directory/leaf_file2
old_file
repeat_name
unusual, name?#
empty_file

